pandas/io/common.py

```diff
@@ -274,64 +273,105 @@ def file_path_to_url(path):
     return urljoin('file:', pathname2url(path))
 
 
-# ZipFile is not a context manager for <= 2.6
-# must be tuple index here since 2.6 doesn't use namedtuple for version_info
-if sys.version_info[1] <= 6:
-    @contextmanager
-    def ZipFile(*args, **kwargs):
-        with closing(zipfile.ZipFile(*args, **kwargs)) as zf:
-            yield zf
-else:
-    ZipFile = zipfile.ZipFile
+def _get_handle(path_or_buf, mode, encoding=None, compression=None,
+                memory_map=False):
+    """
+    Get file handle for given path/buffer and mode.
 
+    Parameters
+    ----------
+    path_or_buf :
+        a path (str) or buffer
+    mode : str
+        mode to open path_or_buf with
+    encoding : str or None
+    compression : str or None
+        Supported compression protocols are gzip, bz2, zip, and xz
+    memory_map : boolean, default False
+        See parsers._parser_params for more information.
 
-def _get_handle(path, mode, encoding=None, compression=None, memory_map=False):
-    """Gets file handle for given path and mode.
+    Returns
+    -------
+    A file like object.
     """
-    if compression is not None:
-        if encoding is not None and not compat.PY3:
-            msg = 'encoding + compression not yet supported in Python 2'
+
+    f = path_or_buf
+    is_path = isinstance(path_or_buf, compat.string_types)
+
+    # in Python 3, convert BytesIO or fileobjects passed with an encoding
+    if compat.PY3 and isinstance(path_or_buf, compat.BytesIO):
+        from io import TextIOWrapper
+        return TextIOWrapper(path_or_buf, encoding=encoding)
+
+    elif compression:
```

Comment:
Currently, this function (`_get_handle`) cannot consume `compression=infer`. However, it would be an easy addition:

```python
if compression == 'infer':
    assert is_path
    compression = infer_compression(path_or_buff)
```

One issue I'm having is that `_get_handle` was undocumented before this PR. So it's a bit hard for me to know what to implement. @jreback or @jorisvandenbossche: can you comment on the tentative docstring and fill in the missing details, so I know exactly what to implement?